[BUGFIX beta] fix regression of clicking link-to with disabled=true

simonihmig opened the pull request:
#15759 introduced a regression, allowing clicking on a `{{#link-to "somewhere" disabled=true}}` to still invoke the transition. This was because the default value of `_isDisabled` was moved to the init hook here: https://github.com/emberjs/ember.js/commit/8c68f5b26f50af456af214065d3b411006d420da#diff-49415576dd428811b5854523cc9b9adcR528, effectively overriding any previously set value. 

The added test covers this, and was failing before the change here.

simonihmig committed:
[BUGFIX beta] fix regression of clicking link-to with disabled=true

bekzod commented:
I think `disabled` is private in `link-to`, and you are supposed to used `disabledWhen` instead and https://github.com/emberjs/ember.js/pull/15780 supposed to fix it

simonihmig commented:
Hm, not sure... 
* even if it's private, it is supposed to work correctly, and all versions of Ember <= 2.17 do support the usage I fixed here, only beta and canary started to [fail](https://travis-ci.org/kaliber5/ember-bootstrap/builds/313485762)...
* the documentation regarding this is actually a bit misleading (cc @locks), given that there is the API docs of the `LinkComponent`, but also the `link-to` helper docs (*which are one and the same thing nowadays, right?*):
  * the [API docs](https://emberjs.com/api/ember/2.17/classes/LinkComponent?show=inherited%2Cprotected%2Cprivate) of the `LinkComponent` mark `disabled` as private, but do not mention `disabledWhen` at all!
  * the `link-to` [helper docs](https://emberjs.com/api/ember/2.17/classes/Ember.Templates.helpers/methods/link-to?anchor=link-to) mention both `disabled` (though private!?) and `disabledWhen`
* given that 👆, even if `disabled` is indeed private and mentioned by mistake in the docs, I would consider it an "intimate API" (that's how you call it, right?), i.e. should only be removed with a prior deprecation

So I would say either remove it all together (no-go IMHO), or fix the bug!?

rwjblue commented:
Agreed. We should fix this. Even if we decide later to deprecate.

simonihmig commented:
Anything missing here? /cc @rwjblue

Commit:
[BUGFIX beta] Refactor `disabled` system in `{{link-to}}`.

The prior system was overly complicated (it still kinda is, but thats a
story for a future refactor). It relied on the fact that `_isDisabled`
was set to disable the link-to (either via hitting the `disabled` CPs
setter, or via passing `disabledWhen`) and then checked `_isDisabled`
when being invoked.

The original issue was that we were _always_ setting `_isDisabled` to
false in the constructor (done to ensure `link-to` has a consistent
shape). That was fixed by pre-sloting `_isDisabled` _before_ calling
`this._super` (yes, I know thats generally bad, but so is our
`Object.assign(instance, properties)` that runs _before_ `init`).

Commit:
[BUGFIX release] Cleanup documentation for `{{link-to}}`.

rwjblue commented:
@simonihmig - Just updated the implementation here, mind taking another look?

simonihmig commented on packages/ember-glimmer/lib/components/link-to.ts:
```diff
@@ -534,10 +518,14 @@ const LinkComponent = EmberComponent.extend({
       return false;
     },
+
     set(_key: string, value: any): boolean {
```
Unrelated to the changes here, but the typing seems wrong. Shouldn't it be `set(...): string|false`?

rwjblue replied:
@simonihmig - Yes, this should be typed as a `string | null` I believe.

simonihmig commented on packages/ember-glimmer/lib/components/link-to.ts:
```diff
@@ -72,25 +72,9 @@
-  any passed value to `disabled` will disable it except `undefined`.
-  to ensure that only `true` disable the `link-to` component you can
-  override the global behavior of `LinkComponent`.
+  any truthy value passed to `disabled` will disable it except `undefined`.
```
"except `undefined`" can be removed I think, as `undefined` is not truthy anyways?

rwjblue replied:
GAH! Thank you

simonihmig commented on packages/ember-glimmer/lib/components/link-to.ts:
```diff
@@ -534,10 +518,14 @@ const LinkComponent = EmberComponent.extend({
+
     set(_key: string, value: any): boolean {
-      if (value !== undefined) { this.set('_isDisabled', value); }
+      this._isDisabled = value;
```
The changes here some right to me in general, but they do change the semantics a bit, don't they (the truthy vs. not undefined)? Could they break something for somebody?

rwjblue replied:
@simonihmig if you take a look at where `this._isDisabled` is actually checked (in `_invoke`), we have been doing a truthy check there for a while so changing this to always update it just makes it simpler...

simonihmig commented:
@rwjblue Uh, already merged? :) I added a few comments...
